pkg/controller/certificaterequests/approver/approver_test.go

```diff
@@ -0,0 +1,240 @@
+				ObjectMeta: metav1.ObjectMeta{Namespace: "testns", Name: "test"},
+				Status: cmapi.CertificateRequestStatus{
+					Conditions: []cmapi.CertificateRequestCondition{
+						{
+							Type:   cmapi.CertificateRequestConditionReady,
+							Status: cmmeta.ConditionFalse,
+							Reason: cmapi.CertificateRequestReasonPending,
+						},
+					},
+				},
+			},
+			expectedConditions: []cmapi.CertificateRequestCondition{
+				{
+					Type:   cmapi.CertificateRequestConditionReady,
+					Status: cmmeta.ConditionFalse,
+					Reason: cmapi.CertificateRequestReasonPending,
+				},
+				{
+					Type:               cmapi.CertificateRequestConditionApproved,
+					Status:             cmmeta.ConditionTrue,
+					Reason:             "cert-manager.io",
+					Message:            ApprovedMessage,
+					LastTransitionTime: &metaNow,
+				},
+			},
+			expectedEvent: "Normal cert-manager.io Certificate request has been approved by cert-manager.io",
+		},
+	}
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			// Create and initialise a new unit test builder
+			builder := &testpkg.Builder{
+				T:     t,
+				Clock: fakeclock.NewFakeClock(now),
+			}
+			if test.request != nil {
+				builder.CertManagerObjects = append(builder.CertManagerObjects, test.request)
+			}
+			builder.Init()
+
+			c := new(Controller)
+			_, _, err := c.Register(builder.Context)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if test.expectedConditions != nil {
+				if test.request == nil {
+					t.Fatal("cannot expect an Update operation if test.request is nil")
+				}
+				expectedRequest := test.request.DeepCopy()
+				expectedRequest.Status.Conditions = test.expectedConditions
+				builder.ExpectedActions = append(builder.ExpectedActions,
+					testpkg.NewAction(coretesting.NewUpdateSubresourceAction(
+						cmapi.SchemeGroupVersion.WithResource("certificaterequests"),
+						"status",
+						test.request.Namespace,
+						expectedRequest,
+					)),
+				)
+			}
+			if test.expectedEvent != "" {
+				builder.ExpectedEvents = []string{test.expectedEvent}
+			}
+			// Start the informers and begin processing updates
+			builder.Start()
+			defer builder.Stop()
+
+			key := test.key
+			if key == "" && test.request != nil {
+				key, err = controllerpkg.KeyFunc(test.request)
+				if err != nil {
+					t.Fatal(err)
+				}
+			}
+
+			// Call ProcessItem
+			err = c.ProcessItem(context.Background(), key)
+			switch {
+			case err != nil:
+				if test.err != err.Error() {
+					t.Errorf("error text did not match, got=%s, exp=%s", err.Error(), test.err)
+				}
+			default:
+				if test.err != "" {
+					t.Errorf("got no error but expected: %s", test.err)
+				}
+			}
+
+			if err := builder.AllEventsCalled(); err != nil {
+				builder.T.Error(err)
+			}
+			if err := builder.AllActionsExecuted(); err != nil {
+				builder.T.Error(err)
+			}
+			if err := builder.AllReactorsCalled(); err != nil {
+				builder.T.Error(err)
+			}
```

Comment:
Nice tests. I ran the tests locally and checked the coverage report, and I can see that these cover both ProcessItem and the Sync method.

Weirdly, the coverage report shows the cases of the switch statement  in Sync as not covered, but I think that's a bug in the coverage tool.

![image](https://user-images.githubusercontent.com/978965/111460898-719ad400-8714-11eb-8523-fa386e0dbd11.png)

Reply:
Yeah, I think that it is a render bug in the coverage reporter. That switch is definitely being hit on all cases in the tests.